Initial Commit For api-02-async-await Objective

MarvelProgramming opened the pull request:
For issue #40

Commit:
Initial Commit for api-02-async-await Objective

MarvelProgramming commented:
Currently, when attempting to use `require` in the QuestIDE, the error "require is not defined" occurs. I'm having trouble pinpointing the exact cause. From what I've gathered, others who've run into this end up changing `contextIsolation` to `false` and `nodeIntegration` to `true` when [instantiating a new BrowserWindow](https://github.com/twilio/twilioquest/blob/3.2_next/src/js/app/common/code_runner.js#:~:text=this.codeWindow%20%3D,%7D\)%3B). In the version of Electron that TQ currently uses, [`false` is the default value for `contextIsolation`](https://github.com/electron/electron/blob/v8.2.4/docs/api/browser-window.md#:~:text=separate%20JavaScript%20context.-,Defaults%20to%20false,-.%20The%20context%20that) and we're already setting `nodeIntegration` to `true`. I've done a once-over on the example and validator, but am blocked on this objective until I can figure out this issue with `require`

\*UPDATE\*: This is taken care of in the "[Added String.replaceAll and Require to Code Contexts](https://github.com/twilio/twilioquest/pull/902)" PR

Commit:
Updated Objective

-Added first draft of description
-Fixed bug in validator where incorrect test directory path was being used
-Updated example.js verbiage, readdir import, and test logging implementation

MarvelProgramming commented on levels/inside_perimeter/objectives/api-02-async-await/example.js:
```diff
@@ -0,0 +1,10 @@
+  // TODO: return the last file in dirPath!
+}
+
+(async function () {
```
Not sure how we feel bout this way of writing a "test run" of the user's function, but I had a hard time thinking of another solution that wouldn't end up being a time sink

rmkubik replied:
I think it would be clearer if we didn't use the IIFE. It's a pretty weird pattern for new devs, I think!

```js
async function runTests() {
  console.log("Test case data:");
  console.log(await findLastFileInDir("."));
  // console.log(await findLastFileInDir("REPLACE_THIS_WITH_FILE_PATH"));
}

runTests();
```

I also added a findLastFileInDir test case with a dir in it to the example so the player will see _some_ output.

Commit:
Merge branch 'milestone-1-aa3' of github.com:TwilioQuest/twilioquest-api-academy into 40-api-02-async-await-learn-async-await-locally

Commit:
Moved Objective To lovelace_tower

Commit:
Merge branch 'milestone-1-aa3' of github.com:TwilioQuest/twilioquest-api-academy into 40-api-02-async-await-learn-async-await-locally

Commit:
Added Walkthrough

joenash commented on levels/lovelace_tower/objectives/api-02-async-await/walkthrough.md:
```diff
@@ -0,0 +1,99 @@
+# Running code asynchronously
+
+Most of the code we write runs instantly\*, but situations exist in which it takes longer than usual. One such situation is when you're reading the contents of a directory using NodeJS and Javascript. In Javascript, we have the ability to execute code asynchronously. That is to say, without stopping everything else to wait for that code to finish.
```
```suggestion
Most of the code we write returns a value instantly\*, but there are times when it takes longer than usual. When this happens, Javascript doesn't wait: it will keep executing the rest of our code, even if that code depends on the value we are waiting to be returned!

An example of this is reading the contents of a directory using NodeJS and Javascript. Take a look at this code:
```

joenash commented on levels/lovelace_tower/objectives/api-02-async-await/walkthrough.md:
````diff
@@ -0,0 +1,99 @@
+console.log(contents);
+```
+
+The example uses `readdir` to get files and folders in a directory. The `readdir` function is asynchronous and will take some time to finish. With things as they are, our `console.log` would print `Promise { <pending> }`. That's where promises come into the picture.
````
```suggestion
The example uses `readdir` to get files and folders in a directory, and assigns those files and folders to a variable `contents`, and then outputs those contents to the console. Because `readdir` takes time to return a value, and Javascript doesn't wait, our `console.log` may not output what we expect.

We call functions like this "asynchronous", and they return a Promise. In this case, our `console.log` would print `Promise { <pending> }`.

The `readdir` function is asynchronous and will take some time to finish. With things as they are, our `console.log` would print `Promise { <pending> }`. That's where promises come into the picture.
```

joenash commented on levels/lovelace_tower/objectives/api-02-async-await/walkthrough.md:
```diff
@@ -0,0 +1,99 @@
+
+## What are promises?
+
+Promises are what asynchronous functions return, and is what Javascript uses to represent the state of asynchronous tasks. They give us the ability to pause our program until a task completes, or provide a callback (i.e. a function) to be invoked once finished. To do so, we use the `await` keyword and `promise.then` method respectively. More on those coming up.
```
```suggestion
Promises represent a value that we will get in the future. Initially, an asynchronous function returns a Promise object, and when the real value is available, the Promise "resolves" and is replaced by the value. In other words, they are a **promise** of a future value! Get it?

Javascript can use the presence of a Promise to control the execution of our program, giving us the ability to pause it until a task completes, or provide a callback (i.e. a function) to be invoked once finished. To do so, we use the `await` keyword and `promise.then` method respectively. More on those coming up.
```

joenash reviewed (commented):
Looking good! Made some changes to the flow and explanation of the first example

MarvelProgramming committed:
Update levels/lovelace_tower/objectives/api-02-async-await/walkthrough.md

Updated walkthrough verbiage

Co-authored-by: Joe  <1790822+joenash@users.noreply.github.com>

MarvelProgramming committed:
Update levels/lovelace_tower/objectives/api-02-async-await/walkthrough.md

Updated walkthrough verbiage

Co-authored-by: Joe  <1790822+joenash@users.noreply.github.com>

MarvelProgramming committed:
Update levels/lovelace_tower/objectives/api-02-async-await/walkthrough.md

Updated walkthrough verbiage

Co-authored-by: Joe  <1790822+joenash@users.noreply.github.com>

nmadadair commented:
In terms of Narrativization here, I think the initial idea was to very simply say something, to paraphrase Ryan, like APIs are like magic, not all spells happen instantly, you sometimes have to wait for return.

Something like:

> _Welcome, Gauntleter, to the second challenge of Lovelace Tower. For this challenge, you must acknowledge the magical powers of asynchronous functions. Because of this asynchronicity, wielders of such API powers must await the return of their functions. But delaying the instant execution of code has a worthwhile power: it can transcend the very concept of linear time._

Commit:
Updated Objective

-Updated description to be more thematic
-Updated example such that it no longer uses an immediately invoked function for running test case(s)

Commit:
Merge branch 'milestone-1-aa3' of github.com:TwilioQuest/twilioquest-api-academy into 40-api-02-async-await-learn-async-await-locally
